src/osd/PG.h

```diff
@@ -625,6 +625,7 @@ class PG : public DoutPrefixProvider,
 
   void on_backfill_reserved() override;
   void on_backfill_canceled() override;
+  void on_recovery_deferred() override {}
```

Comment:
Why is this change only relevant for Crimson? What Classic does differently?
Can you please explain why the previous approach of ignoring already queued request fail to work here?

We should probably try to avoid editing classical's PG or PeeringState if the fix is not relevant to it.

Reply:
Note: There are 2 new PGLog assertions failures I'm looking into:
https://tracker.ceph.com/issues/67219
https://tracker.ceph.com/issues/67366
Most likely they are not related to the fix here but it's worth mentioning.

Reply:
> Note: There are 2 new PGLog assertions failures I'm looking into: https://tracker.ceph.com/issues/67219 https://tracker.ceph.com/issues/67366 Most likely they are not related to the fix here but it's worth mentioning.

Got it:-)

Reply:
> Why is this change only relevant for Crimson? What Classic does differently? Can you please explain why the previous approach of ignoring already queued request fail to work here?
> 
> We should probably try to avoid editing classical's PG or PeeringState if the fix is not relevant to it.

Sorry, I didn't clarify this in the commit message. The case is as follows:

1. PG `P` is under recovery, and `PGRecovery::start_recovery_ops()` is called;
2. PG `P`'s recovery is deferred, and `PGRecovery::start_recovery_ops()` is blocked on waiting for external replies;
3. Before the arrivals of external replies, PG `P` resumes recovery and a new round of `PGRecovery::start_recovery_ops()` starts;
4. The external replies arrives and the old `PGRecovery::start_recovery_ops()` continues as `PG:is_recovering()` returns true, which leads to duplicated recovery ops.

I added this in the commit message, please take a look, thanks:-)

Reply:
> How does Classic avoid this issue? Is it because it's not relying on `PG:is_recovering()` to stop the recovery op in the first place?

It looks that Classic OSDs avoid the duplicated recoveries with the help of both `PG::is_recovering()` and `PG::recovery_queued`, it's only when `PG::is_recovering()` is true and `PG::recovery_queued` is false that a new recovery is queued. Should we change to adopt the same strategy?

Reply:
I would highly prefer using this logic as well if we can. This should help us to avoid PeeringState changes, right?

Reply:
> > How does Classic avoid this issue? Is it because it's not relying on `PG:is_recovering()` to stop the recovery op in the first place?
> 
> It looks that Classic OSDs avoid the duplicated recoveries with the help of both `PG::is_recovering()` and `PG::recovery_queued`, it's only when `PG::is_recovering()` is true and `PG::recovery_queued` is false that a new recovery is queued. Should we change to adopt the same strategy?

@Matan-B Sorry, I was wrong about this, Class OSD doesn't have this issue, it doesn't have counterparts for `BackgroundRecovery`, so it doesn't have to strictly avoid new `PG::start_recovery_ops()` execution when there's inflight recovery ops.

So I think we might need to a new approach for Crimson.